src/ObjectUtils/include/ObjectUtils/PdbDiaUtils.h

```diff
@@ -0,0 +1,352 @@
+  HRESULT result;
+  DWORD tag;
+  if (FAILED(type->get_symTag(&tag))) {
+    ORBIT_ERROR("Found Dia symbol without a tag. Returning \"???\" as a type name.");
+    return "???";
+  }
+
+  std::string result_string;
+  if (tag != SymTagPointerType) {
+    if (BOOL is_const; SUCCEEDED(type->get_constType(&is_const)) && is_const) {
+      result_string += "const ";
+    }
+    if (BOOL is_volatile; SUCCEEDED(type->get_volatileType(&is_volatile)) && is_volatile) {
+      result_string += "volatile ";
+    }
+    if (BOOL is_unaligned; SUCCEEDED(type->get_unalignedType(&is_unaligned)) && is_unaligned) {
+      result_string += "__unaligned ";
+    }
+    if (BOOL is_restricted; SUCCEEDED(type->get_restrictedType(&is_restricted)) && is_restricted) {
+      result_string += "restricted ";
+    }
+  }
+
+  BSTR type_name_bstr = {};
+  std::string type_name;
+  if (SUCCEEDED(type->get_name(&type_name_bstr)) && type_name_bstr != NULL) {
+    std::wstring type_name_wstring{type_name_bstr};
+    type_name = std::string(type_name_wstring.begin(), type_name_wstring.end());
+    SysFreeString(type_name_bstr);
+  }
+
+  if (!type_name.empty()) {
+    result_string += type_name;
+    result_string += parent_pointer_type_str;
+    return result_string;
+  }
+
+  switch (tag) {
+    case SymTagArrayType: {
+      // For now we only print "[]" for arrays. However there is room for improvement.
+      // We could e.g. also print the size of the array if known.
+      IDiaSymbol* base_type = nullptr;
+      if (FAILED(type->get_type(&base_type)) || base_type == nullptr) {
+        return "???";
+      }
+      CComPtr<IDiaSymbol> base_type_comp = base_type;
+      std::string new_pointer_type_str = "[]" + parent_pointer_type_str;
+      result_string += PdbDiaTypeAsString(base_type, new_pointer_type_str);
+
+      return result_string;
+    }
+    case SymTagBaseType: {
+      DWORD base_type;
+      if (FAILED(type->get_baseType(&base_type))) {
+        return "???";
+      }
+      switch (base_type) {
+        case btNoType:  // 0
+          result_string += "<no type>";
+          break;
+        case btVoid:  // 1
+          result_string += "void";
+          break;
+        case btChar:  // 2
+          result_string += "char";
+          break;
+        case btWChar:  // 3
+          result_string += "wchar_t";
+          break;
+        case btInt:  // 6
+        {
+          ULONGLONG length;
+          if (FAILED(type->get_length(&length))) {
+            return "???";
+          }
+          switch (length) {
+            case 1:
+              result_string += "char";
+              break;
+            case 2:
+              result_string += "short";
+              break;
+            case 4:
+              result_string += "int";
+              break;
+            case 8:
+              result_string += "__int64";
+              break;
+            default:
+              ORBIT_UNREACHABLE();
+          }
```

Comment:
That's a switch inside a switch, inside a switch. A little too much nesting for my preference. What about putting some of these switches into free functions or closures? In this case for example it should be rather simple. `GetSignedIntegerTypeFromSizeInBytes` should work (just an example) and it's even repeated below as well!

Reply:
The whole SymTagBaseType case should probably also go into its own free function. The scope is really long.